File: 04-Data-Structures/code/typescript/src/bitset.ts
import { DSError } from "./errors.js";

/** Fixed-size compact boolean array backed by one byte per bit (lab clarity over packing density). */
export class Bitset {
  private bits: Uint8Array;
  private n: number;

  constructor(nBits: number) {
    if (nBits < 0) throw new DSError("invalid", "nBits must be >= 0");
    this.n = nBits;
    this.bits = new Uint8Array(nBits);
  }

  set(i: number, value: boolean = true): void {
    this.checkIndex(i);
    this.bits[i] = value ? 1 : 0;
  }

  get(i: number): boolean {
    this.checkIndex(i);
    return this.bits[i] === 1;
  }

  count(): number {
    let c = 0;
    for (let i = 0; i < this.n; i++) c += this.bits[i];
    return c;
  }

  toBits(): string {
    return Array.from(this.bits).join("");
  }

  private checkIndex(i: number): void {
    if (i < 0 || i >= this.n) throw new DSError("index", `index ${i} out of bounds`);
  }

  checkInvariants(): void {
    if (this.bits.length !== this.n) throw new DSError("invalid", "backing store length mismatch");
    for (let i = 0; i < this.n; i++) {
      if (this.bits[i] !== 0 && this.bits[i] !== 1) throw new DSError("invalid", "non-boolean bit value");
    }
  }
}

File: 04-Data-Structures/code/python/seb_ds/bitset.py
"""Fixed-size compact boolean array. Mirrors `typescript/src/bitset.ts`."""

from __future__ import annotations

from .errors import DSError


class Bitset:
    def __init__(self, n_bits: int) -> None:
        if n_bits < 0:
            raise DSError("invalid", "nBits must be >= 0")
        self._n = n_bits
        self._bits = bytearray(n_bits)

    def set(self, i: int, value: bool = True) -> None:
        self._check_index(i)
        self._bits[i] = 1 if value else 0

    def get(self, i: int) -> bool:
        self._check_index(i)
        return self._bits[i] == 1

    def count(self) -> int:
        return sum(self._bits[: self._n])

    def to_bits(self) -> str:
        return "".join(str(b) for b in self._bits)

    def _check_index(self, i: int) -> None:
        if i < 0 or i >= self._n:
            raise DSError("index", f"index {i} out of bounds")

    def check_invariants(self) -> None:
        if len(self._bits) != self._n:
            raise DSError("invalid", "backing store length mismatch")
        for i in range(self._n):
            if self._bits[i] not in (0, 1):
                raise DSError("invalid", "non-boolean bit value")
